### DISCUSSION TITLE
[Feature] Choose Libraries for Partner Sharing

### DISCUSSION THREAD

**Discussion by dpantel** [immich-app/immich#4500] [Feature Request]
### The feature

Currently Partner Sharing is all-or-nothing. I would like to enable Partner Sharing for some Libraries, not not for others.

For example, I have childhood photos (as an external library) that I want to view on my account, but don't want to share with my partner. But I _do_ want to share my default library and another external library with said partner.

Thank you.

### Platform

- [X] Server
- [ ] Web
- [ ] Mobile

**Comment by v411e** [comment]
Yes, it would be cool to have an option to whitelist or blacklist specific folder paths on sharings between users. Maybe even allowing to share different paths with different users (The photos I want to share with my dad are not the same that I want to share with my partner)

**Reply by amacal1** [reply]
Yes, this is how I imagine it should be. I have NAS with 20+ years of photos and videos (somewhat well organized by event and overlap with other family members). This would make it really easy to make external libraries out of different folders from this storage area. I could assign some of these libraries to all users, but some of them more selectively. More importantly, it shouldn't have to transcode videos and generate thumbnails for each user. That would be enormously taxing on the Immich storage just for thumbnails and video transcodes, but also computationally expensive to transcode each video multiple times.

**Comment by kaaku3** [comment]
I would love partner share based on facial recognition, that's what I did in google photos.

**Comment by michaelhthomas** [comment]
I'd personally like to see libraries exist outside of users, something like how Jellyfin works with media. This might be too radical, but I'm imagining each user having their "default" library where uploads are placed, as well as access to external libraries that they have been added to. The admin will be able to control which external libraries exist, what their paths are, and who they are shared with.

I feel like that would make a bit more sense than the current approach, as having individual users make their own external libraries feels pretty strange, and I think this makes more sense as an administrative task. As a plus, with support for removing assets from external libraries coming, this could help close the gap between "internal" libraries and "internal" libraries.

Here's a quick mockup of what I'm envisioning:
![Mockup of library settings options](https://github.com/immich-app/immich/assets/18223295/54c2cdb1-803e-4fcf-92a0-828acb3112cc)

**Reply by etnoy** [reply]
I love this idea

**Reply by github-cli** [reply]
yea, would love to see this as well!
we have six users in the family, all access to the same huge external library

**Reply by fonz-zz** [reply]
This sounds great! I just imported our external librabry which took a week and then a few more days browsing through the people to hide unknowns from the crowd, give the ones we know a name and birthday and fixing some mismatches.
Imagine my surpsrise when I set up the wifes account and I have to go throught that painfull process again :-( ....
The option of having an external library centraly managed by the admin would be so much time saved!

**Reply by ttyridal** [reply]
Would love to see this too! Although, if the faces/other-meta data sharing (and searching!) were implemented, I think it could be solved by with some additional shared-asset-owner accounts.. 

Could be messy if you have a lot of different people sharing different subsets of the collection though.

**Reply by mrkaryo** [reply]
This would be indeed great to have!!

**Reply by tuupola** [reply]
I think this is the way most Immich users would prefer libraries and sharing to work. Immich is currently built with mindset of setting up a photo hosting saas. With saas it makes sense to have strict privacy settings such as not sharing face recognition data between users.

My gut feeling says most of Immich users are selfhosters who use it for family and friend photos. Sharing photos and working together with someone for example to tag peoples faces is currently too difficult or even impossible.

**Reply by OlJohnny** [reply]
This would be extremely great, useful and would definitly set immich apart from the rest!
Might also make the mental transition easier for self-hosters coming from Jellyfin et al
Probably entails a massive rework of basically all systems.
Could someone drop a hint if this is on any roadmap? 😉

**Reply by amacal1** [reply]
Yes, this is how I imagine it should be. I have NAS with 20+ years of photos and videos (somewhat well organized by event and overlap with other family members). This would make it really easy to make external libraries out of different folders from this storage area. I could assign some of these libraries to all users, but some of them more selectively. More importantly, it shouldn't have to transcode videos and generate thumbnails for each user. That would be enormously taxing on the Immich storage just for thumbnails and video transcodes, but also computationally expensive to transcode each video multiple times.

**Reply by helderjgoncalves** [reply]
Love this view!! Implement this please!!

**Comment by aisbergde** [comment]
partner sharing in the current way is useless for me, also the partner sharing in Google photos. we don't use it, because it is not possible to control and limit the content. That's why we are using shared albums in our family. Smart albums, based on persons, are great. I think, they are planned also in immich.

**Comment by OrbitingOcelot** [comment]
Hey guys, after a lot of fiddling around I sadly found out that I have to import our external library four times. Our use case is we have four users, each accesses a very large shared base library (added as external library), and then everyone has their own stuff on top of it. Is there any better way to currently achieve what I want without importing the same thing four times?

Would be really great to have access control on shared library level, this sounds like a clean and neat solution. Any idea if this might appear on the roadmap some day?

**Comment by kaaku3** [comment]
I would setup 5 users, 4 actual and one shared. Use partner sharing to share all the pics from the extra account.

**Reply by OrbitingOcelot** [reply]
Oh, that's a quite clever workaround, I like it, thank you! The only problem is that the partner shared files are not shown in the main timeline, at least I could not find a way to make them appear... 😣

**Reply by ahayzen** [reply]
You can show them in your timeline by using the timeline integration https://immich.app/docs/features/partner-sharing/#timeline-integration

This almost works as a workaround until you can share a library with multiple users but there are a few remaining issues

- You can create albums from shared partner photos from the timeline, but you can't select from the add photos of the album #6486 
- Facial recognition data is not shared from the partner photos #5089

**Reply by OrbitingOcelot** [reply]
Thanks @kaaku3 and @ahayzen - almost there!

Now the last missing feature for a full integration would be to display the "partner sharing" pictures on the map, that does not seem to work yet - unless you guys also have an idea regarding this?

**Reply by ttyridal** [reply]
> Now the last missing feature for a full integration would be to display the "partner sharing" pictures on the map, that does not seem to work yet - unless you guys also have an idea regarding this?

That's already implemented :) go to options on the map and "show partners photos"

**Reply by OrbitingOcelot** [reply]
> That's already implemented :) go to options on the map and "show partners photos"

Oh wow now I feel really stupid :-) Thank you!

Though I have to say, the small cog button on the map was quite unexpected, would be nice to have this also show up in the settings menu, below "Partner Sharing" -> "Show in timeline", would probably be easier to find...

**Reply by phileastv** [reply]
This workaround works great ! But have you an idea on how to show partner medias on the "Folder" view ? I think the main use of external libraries is for pictures of "legacy" libraries, which are organized in a folder manner (like old trips pictures...), so it makes sense to see theses medias on a folder view, for multiple people.

**Comment by ttyridal** [comment]
Totally agree! but this was the judgment of the project when I implemented
that one ;)

**Comment by wonx** [comment]
I'd also be interested in this. We have a main shared picture library with all family members, and then each one has its own private library. Right now the best workaround seems to create a separate user for the shared library, but it definitely feels like a workaround. Some kind of per user access permissions on each library would be idea, like the mockup that @michaelhthomas provided.

**Comment by andyghc** [comment]
Ditto here.  This is a feature that would be desirable.  

I think this can be worked around by setting up a shared account that we both use, and I'm sure that will work sufficiently well but also has its own draw-backs.

**Comment by sirpaddy** [comment]
I really like Immich so far, but this is one thing I miss. I guess this would be a quite common scenario, that you would like to share whole libraries, but not ALL photos like in partner sharing. Would appreciate, if something is coming in the future:)
Thank you!

**Comment by gschintgen** [comment]
See also: #12614 (sharing in general).

**Comment by enlight3d** [comment]
Hello, really interested in this feature as well ! Partner sharing kinda works for me, but what I'm missing is the folder view. Partner can't see the folder structure of the shared library 😞

**Reply by etnoy** [reply]
Did you enable folder view for the other account?

**Reply by enlight3d** [reply]
Yup. On my account I see the folder structure, but on the second one, it is empty.

**Reply by olegartys** [reply]
Hi, empty folder view in partner account looks like a bug, is there a separate issue for this?

**Comment by sol3uk** [comment]
+1 on this, I'm setting up immich with me and my brother and in our use case we want somewhere we can both dump our photos to preview and review/edit/delete together, to be used by either one of us, family style, but I don't want to have to give him access to all of my photos, he's not my partner...
The alternative is we setup one folder for each of us as an external library, this is fine, but again it places an unnecessary extra step between having access to the images straight away, we would need to manually take a step and either add them to a shared album, or otherwise manually share those images with eachother.
A shared external library would remove this need and allow us to predictably go through our workflows without locking eachother out of seeing any images we know will always be shared anyway.

**Comment by zilexa** [comment]
+1 for library-specific partner sharing please!
We want 1 library (could be the External Library!) fully shared, all rights for both.
And then each have our own libraries.

**Comment by greentown0** [comment]
+1. Current partner sharing is all or nothing. A bit of granularity is needed to accommodate many more use cases.

**Comment by mrkaryo** [comment]
I think if it was like the icloud’s scheme, a per-photo granularity, it would be the best!

**Reply by LMSSonos** [reply]
My usecase is that I have a External Library, this includes "cleanedup" fotos of the family. Would be awesome to share that with multiple user. Permission-Wise it would be sufficient to have `Admin` and `ViewOnly` permissions and not on per-photo level ;-)

**Comment by jamabaiz** [comment]
I am new in Immich, but lack this feature from Synology Photos. For "familly" shared library and "personal" private library.
Not entered so deep in source-code yet, but I don't see how it would work for "libraries", by design, but...

Assuming that users have folders/separate drives/mounts like:
* /mnt/family/photos/
* /mnt/personal/photos/

I wonder if in current state of "Immich". Wouldn't it be easy to create script. Which will process library `/mnt/familly/`. Scan paths, and for each directory, make album containing all photos. Adding then all users to that created albums. So Album-Per-Directory, everything should be able done by existing API. It would be easy work-around, until it will be somehow natively supported.

The only consideration is how to handle sync, but should be so hard. Assuming that this script on creation will store ids of albums, and immich photo id per photo path, it will can add missing albums/photos, by diffs "local eg. sqlite db" and "immich api".

Let me know if somebody would be interested in such _solution_.

**Comment by aomanchuria** [comment]
I finally made the move from photoprism to immich. Things were looking good as my phone photos were showing up and then all my DSLR shared libraries too...with all 15 years of kid photos. Then I installed it on my wife's phone thinking of expecting the same libraries there. Nope. I hit the wall a few times with my head, then I realized that I could create the same external libraries separately for her. But nope, that didn't populate the photos.

I'm an avid photographer and take hundreds of photos. Good or bad I never delete. My wife on the other hand has zero heart for artistic bokeh and motion blur. She has no eye for grant black and white images from totally planned under or over exposed shots. So I want her to see all our family photos and be able to create her own view of the same shared libraries.

Good to see this issue hasn't been closed. And wow! What a difference 2 years can make. I'm super impressed with the latest version.

**Reply by amacal1** [reply]
The short term solution for this is to create a user named "Family Photos" in addition to a user account for you and a third user account for your wife. Assign the external libraries with family photos to the Family Photos user. Finally, you "partner share" the assets from Family Photos to your wife and your user accounts. Family Photos 'owns' the family photos and you and your wife can see them because they're shared with you. 

A neat thing is that each of you can toggle them on and off for things like the maps view and the library. The really crappy part is that you can't search those photos using the object/face detection. So, it's not perfect. But, it's better than nothing. I do hope they can fix this and make shared assets searchable, or - better yet - make sharing external libraries directly shareable.

**Reply by navin2004** [reply]
> The short term solution for this is to create a user named "Family Photos" in addition to a user account for you and a third user account for your wife. Assign the external libraries with family photos to the Family Photos user. Finally, you "partner share" the assets from Family Photos to your wife and your user accounts. Family Photos 'owns' the family photos and you and your wife can see them because they're shared with you.
> 
> A neat thing is that each of you can toggle them on and off for things like the maps view and the library. The really crappy part is that you can't search those photos using the object/face detection. So, it's not perfect. But, it's better than nothing. I do hope they can fix this and make shared assets searchable, or - better yet - make sharing external libraries directly shareable.

I hit the same issue after setting up everything nicely, added my wife and she cannot see anything. I created the Family Photos user and added that user under Partner sharing as well as my wife. But it won't allow face data. What did you mean by "Assign the external libraries with family photos to the Family Photos use" ? Do you mean make the Family photos user as Admin account ? I don't see any option to do this part.

**Comment by mselley** [comment]
Is this possible yet or any workarounds? 
It would make it Immich perfect for sharing family photos. Currently my main struggle is this.

**Reply by ttyridal** [reply]
You can configure a separate user that owns the libarary and then partner-share from that account. You will "loose" (not able to see) all metadata though.

the alternative is forks, with all the related disaventages. Seems everybody has one these days (me to! https://github.com/ttyridal/fimmich)

**Comment by photoserver-pro** [comment]
This request is closely related to a user story I've described in the Better Sharing thread, just from a slightly different angle.

The underlying question is the same: the external library should be the natural unit of sharing — not the individual asset, not the album. Which users have access to which library, and with what permissions, is a much cleaner model than the current all-or-nothing approach.
https://github.com/immich-app/immich/issues/12614#issuecomment-4106427277


Hey @mselley  For those interested in seeing where library-level sharing could go technically, the open-noodle/gallery fork is an interesting case study — it's exploring exactly this kind of model in practice: https://github.com/open-noodle/gallery/discussions/133 & https://github.com/immich-app/immich/issues/12614#issuecomment-4108674385
@deeds67

**Comment by peca89** [comment]
+1 eagerly waiting on external library sharing feature

**Comment by tsenwrath** [comment]
This is exactly what I need. My wife and I each back up our phones to separate Immich accounts and want Partner Sharing for those phone uploads, but we each also have private external libraries containing older personal photos. I need to share my upload/default library without automatically sharing every external library attached to my account.
